File: frontend/src/services/extensions.ts
/**
 * 扩展管理 API 封装
 */

import { apiFetch } from './http';
import type { ListResponse } from './types';

export type ExtensionTransport = 'stdio' | 'http';
export type ExtensionStatus = 'enabled' | 'disabled';
export type ExtensionHttpProtocol = 'streamable_http';
export type ExtensionAuthType = 'none' | 'bearer' | 'basic';
export type ExtensionConnectionStatus = 'ok' | 'degraded' | 'failed';

export interface HttpAuthConfig {
  type: ExtensionAuthType;
  token?: string | null;
}

export interface ExtensionHttpConfig {
  url: string;
  protocol: ExtensionHttpProtocol;
  headers: Record<string, string>;
  auth: HttpAuthConfig;
  timeout_seconds?: number | null;
}

export interface ExtensionStdioConfig {
  template_id: string;
  args: string[];
  env: Record<string, string>;
  timeout_seconds?: number | null;
}

export interface ExtensionSecurityConfig {
  allowlist_tools: string[];
  confirmation_required: boolean;
}

export interface ExtensionObservabilityConfig {
  emit_metrics: boolean;
  log_level_override?: 'DEBUG' | 'INFO' | 'WARNING' | 'ERROR' | null;
}

export interface ToolExtension {
  id: string;
  name: string;
  transport: ExtensionTransport;
  status: ExtensionStatus;
  http_config: ExtensionHttpConfig | null;
  stdio_config: ExtensionStdioConfig | null;
  security_config: ExtensionSecurityConfig;
  observability_config: ExtensionObservabilityConfig | null;
  created_at: string;
  updated_at: string;
}

export interface ToolExtensionCreate {
  name: string;
  transport: ExtensionTransport;
  status?: ExtensionStatus;
  http_config?: ExtensionHttpConfig | null;
  stdio_config?: ExtensionStdioConfig | null;
  security_config: ExtensionSecurityConfig;
  observability_config?: ExtensionObservabilityConfig | null;
}

export interface ToolExtensionUpdate {
  name?: string;
  transport?: ExtensionTransport;
  status?: ExtensionStatus;
  http_config?: ExtensionHttpConfig | null;
  stdio_config?: ExtensionStdioConfig | null;
  security_config?: ExtensionSecurityConfig;
  observability_config?: ExtensionObservabilityConfig | null;
}

export interface ToolDescriptor {
  name: string;
  description: string | null;
  input_schema: Record<string, unknown> | null;
}

export interface ToolDescriptorListResponse extends ListResponse<ToolDescriptor> {
  connection_status: ExtensionConnectionStatus;
  last_error: string | null;
  latency_ms: number | null;
}

export interface StdioTemplateDescriptor {
  id: string;
  label: string;
  description: string | null;
  command: string;
  args: string[];
}

export interface StdioTemplateListResponse {
  items: StdioTemplateDescriptor[];
}

export type ToolExtensionListResponse = ListResponse<ToolExtension>;

/**
 * 获取扩展列表
 */
export async function listExtensions(params?: {
  status_filter?: ExtensionStatus;
  skip?: number;
  limit?: number;
}): Promise<ToolExtensionListResponse> {
  const searchParams = new URLSearchParams();
  if (params?.status_filter) searchParams.set('status_filter', params.status_filter);
  if (params?.skip !== undefined) searchParams.set('skip', String(params.skip));
  if (params?.limit !== undefined) searchParams.set('limit', String(params.limit));
  const query = searchParams.toString();
  return apiFetch<ToolExtensionListResponse>(`/api/v1/extensions${query ? `?${query}` : ''}`);
}

/**
 * 获取扩展详情
 */
export async function getExtension(extensionId: string): Promise<ToolExtension> {
  return apiFetch<ToolExtension>(`/api/v1/extensions/${extensionId}`);
}

/**
 * 获取 STDIO 模板列表
 */
export async function listStdioTemplates(): Promise<StdioTemplateListResponse> {
  return apiFetch<StdioTemplateListResponse>('/api/v1/extensions/stdio-templates');
}

/**
 * 创建扩展
 */
export async function createExtension(data: ToolExtensionCreate): Promise<ToolExtension> {
  return apiFetch<ToolExtension>('/api/v1/extensions', {
    method: 'POST',
    body: JSON.stringify(data),
  });
}

/**
 * 更新扩展
 */
export async function updateExtension(
  extensionId: string,
  data: ToolExtensionUpdate
): Promise<ToolExtension> {
  return apiFetch<ToolExtension>(`/api/v1/extensions/${extensionId}`, {
    method: 'PATCH',
    body: JSON.stringify(data),
  });
}

/**
 * 删除扩展
 */
export async function deleteExtension(extensionId: string): Promise<void> {
  await apiFetch<void>(`/api/v1/extensions/${extensionId}`, {
    method: 'DELETE',
  });
}

/**
 * 获取扩展工具列表
 */
export async function getExtensionTools(
  extensionId: string,
  params?: { skip?: number; limit?: number }
): Promise<ToolDescriptorListResponse> {
  const searchParams = new URLSearchParams();
  if (params?.skip !== undefined) searchParams.set('skip', String(params.skip));
  if (params?.limit !== undefined) searchParams.set('limit', String(params.limit));
  const query = searchParams.toString();
  return apiFetch<ToolDescriptorListResponse>(
    `/api/v1/extensions/${extensionId}/tools${query ? `?${query}` : ''}`
  );
}

File: backend/src/app/schemas/extensions.py
"""扩展管理相关 Schemas。"""

from __future__ import annotations

import uuid
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.schemas.pagination import PageMeta


class ExtensionTransport(str, Enum):
    STDIO = "stdio"
    HTTP = "http"


class ExtensionStatus(str, Enum):
    ENABLED = "enabled"
    DISABLED = "disabled"


class ExtensionHttpProtocol(str, Enum):
    STREAMABLE_HTTP = "streamable_http"


class ExtensionAuthType(str, Enum):
    NONE = "none"
    BEARER = "bearer"
    BASIC = "basic"


class ExtensionConnectionStatus(str, Enum):
    OK = "ok"
    DEGRADED = "degraded"
    FAILED = "failed"


class HttpAuthConfig(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: ExtensionAuthType = ExtensionAuthType.NONE
    token: str | None = Field(None, min_length=1, max_length=4096)

    @model_validator(mode="after")
    def _validate_auth(self) -> "HttpAuthConfig":
        if self.type == ExtensionAuthType.NONE:
            return self
        if not self.token:
            raise ValueError("auth.token is required for bearer/basic authentication")
        return self


class ExtensionHttpConfig(BaseModel):
    model_config = ConfigDict(extra="forbid")

    url: str = Field(..., min_length=1, max_length=2048)
    protocol: ExtensionHttpProtocol = ExtensionHttpProtocol.STREAMABLE_HTTP
    headers: dict[str, str] = Field(default_factory=dict)
    auth: HttpAuthConfig = Field(
        default_factory=lambda: HttpAuthConfig(
            type=ExtensionAuthType.NONE,
            token=None,
        )
    )
    timeout_seconds: int | None = Field(None, ge=1, le=600)

    @model_validator(mode="after")
    def _validate_url(self) -> "ExtensionHttpConfig":
        url = self.url.strip().lower()
        if not url.startswith(("http://", "https://")):
            raise ValueError("http_config.url must start with http:// or https://")
        return self


class ExtensionStdioConfig(BaseModel):
    model_config = ConfigDict(extra="forbid")

    command: str = Field(..., min_length=1, max_length=512)
    args: list[str] = Field(default_factory=list, max_length=64)
    env: dict[str, str] = Field(default_factory=dict)
    cwd: str | None = Field(None, min_length=1, max_length=2048)
    timeout_seconds: int | None = Field(None, ge=1, le=600)

    @model_validator(mode="after")
    def _validate_command(self) -> "ExtensionStdioConfig":
        command = self.command.strip()
        if not command:
            raise ValueError("stdio_config.command is required")
        self.command = command
        if self.cwd is not None:
            cwd = self.cwd.strip()
            self.cwd = cwd or None
        return self


class ToolExtensionCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(..., min_length=1, max_length=128)
    transport: ExtensionTransport
    status: ExtensionStatus = ExtensionStatus.ENABLED
    http_config: ExtensionHttpConfig | None = None
    stdio_config: ExtensionStdioConfig | None = None

    @model_validator(mode="after")
    def _validate_transport_config(self) -> "ToolExtensionCreate":
        if self.transport == ExtensionTransport.HTTP and self.http_config is None:
            raise ValueError("http_config is required when transport=http")
        if self.transport == ExtensionTransport.STDIO and self.stdio_config is None:
            raise ValueError("stdio_config is required when transport=stdio")
        if self.transport == ExtensionTransport.HTTP:
            self.stdio_config = None
        if self.transport == ExtensionTransport.STDIO:
            self.http_config = None
        return self


class ToolExtensionUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(None, min_length=1, max_length=128)
    transport: ExtensionTransport | None = None
    status: ExtensionStatus | None = None
    http_config: ExtensionHttpConfig | None = None
    stdio_config: ExtensionStdioConfig | None = None


class ToolExtensionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    transport: ExtensionTransport
    status: ExtensionStatus
    http_config: ExtensionHttpConfig | None = None
    stdio_config: ExtensionStdioConfig | None = None
    created_at: datetime
    updated_at: datetime


class ToolDescriptor(BaseModel):
    name: str
    description: str | None = None
    input_schema: dict | None = None


class ToolExtensionListResponse(BaseModel):
    """扩展列表响应。"""

    items: list[ToolExtensionRead]
    page: PageMeta


class ToolDescriptorListResponse(BaseModel):
    """扩展工具列表响应。"""

    items: list[ToolDescriptor]
    page: PageMeta
    connection_status: ExtensionConnectionStatus = ExtensionConnectionStatus.OK
    last_error: str | None = None
    latency_ms: int | None = None
